Colorbar only tut

patniharshit opened the pull request:
<!--Thank you so much for your PR! To help us review, fill out the form
to the best of your ability.  Please make use of the development guide at
https://matplotlib.org/devdocs/devel/index.html-->

<!--Provide a general summary of your changes in the title above, for
example "Raises ValueError on Non-Numeric Input to set_xlim".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

<!--If you are able to do so, please do not create the
PR out of master, but out of a separate branch.  See
https://matplotlib.org/devel/gitwash/development_workflow.html for
instructions.-->

## PR Summary

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->
Trying to convert colorbar_only example to tutorial.

cc @choldgraf 

<!--If it fixes an open issue, please link to the issue here.-->
#8561 

## PR Checklist

- [ ] Has Pytest style unit tests
- [x] Code is PEP 8 compliant 
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/whats_new.rst if major new feature
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or the
recommended next step seems overly demanding , or if you would like help in
addressing a reviewer's comments.  And please ping us if you've been waiting
too long to hear back on your PR.-->

story645 commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,112 @@
+import matplotlib.pyplot as plt
+import matplotlib as mpl
+
+fig = plt.figure(figsize=(8, 3))
```
I wonder if you can get away with this since axes stuff isn't the point of this tutorial:
`fig, (ax1, ax2, ax3) = plt.subplots(nrows=3, figsize=(8,3))`

patniharshit replied:
Even figsize is not necessary so removed that too.

story645 commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,112 @@
+# ---------------------------
+#
+# The second example illustrates the use of a ListedColormap which generates
+# colormap from a set of listed colors, a BoundaryNorm which generates a
```
generates a

patniharshit replied:
Done.

story645 commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,112 @@
+#
+# The second example illustrates the use of a ListedColormap which generates
+# colormap from a set of listed colors, a BoundaryNorm which generates a
+# colormap index based on discrete interval and extended ends to show the
```
discrete intervals

patniharshit replied:
Done.

story645 commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,112 @@
+# Colorbar with custom extension lengths
+# --------------------------------------
+#
+# Now in the third example we illustrate the use of custom length colorbar
```
Here we illustrate the use of custom length colorbar extensions. (They're used on both discrete and continuous colorbars.)

patniharshit replied:
I am guessing here that 'Now in the third example' is needed to be replaced by 'Here'. Correct me if I am wrong.

story645 commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,112 @@
+# Now in the third example we illustrate the use of custom length colorbar
+# extensions, used on a colorbar with discrete intervals. Here we pass colors
+# as RGB triplet. To make the length of each extension the same as the length
+# of the interior colors pass extendfrac argument as auto
```
pass the extendfrac argument

patniharshit replied:
Done.

story645 commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,112 @@
+# ====================
+#
+# `matplotlib.colorbar.ColorbarBase` derives from `ScalarMappable` and puts a
+# colorbar in specified axes, it is the base class with standalone colorbar
```
I think the original is clearer:
 > ColorbarBase derives from ScalarMappable and puts a colorbar in a specified axes, so it has everything needed for a standalone colorbar.

patniharshit replied:
Changed.

story645 commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,112 @@
+#
+# `matplotlib.colorbar.ColorbarBase` derives from `ScalarMappable` and puts a
+# colorbar in specified axes, it is the base class with standalone colorbar
+# drawing functionality. It can be used as-is to make a colorbar for a given
```
no dash in as is

patniharshit replied:
Dash removed.

patniharshit commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,112 @@
+Customized Colorbars Tutorial
+=============================
+
+This tutorial shows how to build colorbars without an attached mappable.
```
Mappable replaced by plot.

patniharshit commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,108 @@
+# the length of each extension the same as the length of the interior colors
+# pass the extendfrac argument as auto.
+
+cmap = mpl.colors.ListedColormap([[0., .4, 1.], [0., .8, 1.],
```
Instead of passing color in RGB format, I feel it will be more clear if colors are referenced by name?

story645 replied:
Totally agree.

patniharshit replied:
What is the name of the color with RGB = (0, 0.4, 1)?

choldgraf commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,108 @@
+
+"""
+
+###############################################################################
```
not a big deal but you technically don't need to be commenting out lines here, just extend the `"""` up until the first bit of code

patniharshit replied:
Changed so as to have uniform format in all tutorials.

choldgraf commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,108 @@
+# object like an image. In this tutorial we will explore what can be done with
+# standalone colorbar.
+#
+# We will start by making a figure of desired size and adding thress axes.
```
`three` axes

choldgraf replied:
Is `ScalarMappable` something we can link to via `:mod:`? If so we should do it, it's going to be unclear to most people what this is (if it's important)

patniharshit replied:
Done.

choldgraf commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,108 @@
+# will be used. Then create the colorbar by calling `ColorbarBase` and
+# specify axis, colormap, norm and orientation as parameters. Here we create
+# a basic continuous colorbar with ticks and labels. There are many more kwargs
+# which can be used to further modify the colorbar.
```
In these kinds of sentences I like to say something like "for more examples, see <link-to-colorbar-examples-if-they-exist>". Something worth thinking of inserting where reasonable!

patniharshit replied:
I wasn't sure where could it point to so gave a link to colorbar api.

choldgraf commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,108 @@
+# outside of the normalized [0,1] range. Here we pass colors as gray shades as
+# a string encoding a float in the 0-1 range.
+#
+# If a ListedColormap is used, the length of the bounds array must be
```
`:class:matplotlib.path.to.ListedColormap`

choldgraf replied:
(and elsewhere in any point that we refer to a specific class / function / module in MPL)

patniharshit replied:
Done.

choldgraf commented on tutorials/colors/colorbar_only.py:
```diff
@@ -0,0 +1,108 @@
+cb3.set_label('Custom extension lengths, some other units')
+
+plt.tight_layout()
+plt.show()
```
stupid question but is there an extra line at the bottom? I think that's best practices but not sure what MPL does as a standard.

patniharshit replied:
Thanks for pointing it out. My text editor was inserting a new line on save, fixed now.

QuLogic replied:
There _should_ be a newline at the end. If GitHub has that little red symbol on the last line, it means there isn't one.

choldgraf replied:
oops - yes I meant we *wanted* the new line, sorry about the confusion

patniharshit replied:
Fixed now.

choldgraf commented:
@patniharshit looking really nice! thanks for putting this together. I've got a couple quick points in there but I think it's looking great.

choldgraf commented on tutorials/colors/colorbar_only.py:
```diff
@@ -83,16 +84,16 @@
-# the length of each extension the same as the length of the interior colors
-# pass the extendfrac argument as auto.
+# colorbar with discrete intervals. To make the length of each extension the
+# same as the length of the interior colors pass the *extendfrac* argument as
```
"...length of the interior colors, use ``extendfrac='auto'``"

choldgraf commented:
small point in there about using double backticks for the kwarg but other than that LGTM

choldgraf commented on tutorials/colors/colorbar_only.py:
```diff
@@ -84,9 +84,8 @@
-# same as the length of the interior colors pass the *extendfrac* argument as
-# *auto*.
+# colorbar with discrete intervals. To make the length of each extension same
+# as the length of the interior colors, use extendfrac='auto'.
```
sorry to be picky, but I meant adding in double backticks around it, so:

```
``extendfrac='auto'``
```

I'd edit the branch if I could but I don't have commit rights :-/

patniharshit replied:
Fixed.

choldgraf commented on tutorials/colors/colorbar_only.py:
```diff
@@ -85,7 +85,7 @@
 # Here we illustrate the use of custom length colorbar extensions, used on a
 # colorbar with discrete intervals. To make the length of each extension same
-# as the length of the interior colors, use extendfrac='auto'.
+# as the length of the interior colors, use ``extendfrac='auto'``.
```
actually do you think you could make this triple backticks?

...JK, this PR looks good to me :)

patniharshit commented:
Whats the issue here? Why check is failing?

choldgraf commented:
Check out the logs in travis: https://travis-ci.org/matplotlib/matplotlib/jobs/231794760

One drawback of sphinx-gallery is that it collects the figure each time it hits another RST block, so you need to re-do `fig, ax = plt.subplots` each time you want a new figure to show up in the rST. It's one of those annoying errors that works fine when you run the example on your own, but breaks with sphinx-gallery.

choldgraf commented:
Looks like tests are passing. @story645 do you know if there's a build that generates the website for us? E.g., MNE uses `circle.ci` to generate the documentation as a part of the tests, which is a really nice sanity check to make sure it looks correct at the end of documentation PRs.

patniharshit commented:
Just removed a line which read like draw a figure and three axes. It lost its value after we switched to making seperate fig and axis in each rst block instead of making three axes initially.

story645 commented:
Sorry @choldgraf dunno. I know at least one of the tests builds docs, but dunno which one it renders. Maybe @tacaswell does?

QuLogic commented:
There are two builds that run the docs, but unfortunately do not upload them anywhere. Travis doesn't hold artifacts though it can go to S3, but I think no-one remembers the key for it...

choldgraf commented:
@QuLogic I've found it really useful for building docs, particularly when building them locally is non-trivial (as it is with MPL). I can ask the MNE dev that set it up for us if they can explain how tough it was, in case it'd be useful for MPL...lemme know!

Otherwise, this PR looks good to me
